### ISSUE TITLE
sql: get rid of upVersion bit on TableDescriptor

### ISSUE THREAD

**Issue by vivekmenezes** [cockroachdb/cockroach#25030]
TableDescriptor has an up_version bit . From a comment in structured.proto

// Monotonically increasing version of the table descriptor.
  //
  // Invariants:
  // 1. not more than two subsequent versions of the table
  // descriptor can be leased. This is to make the system
  // easy to reason about, by permitting mutation state
  // changes (reflected in the next version), only when the existing
  // state (reflected in the current version) is present on all
  // outstanding unexpired leases.
  // 2. A schema change command (ALTER, RENAME, etc) never directly
  // increments the version. This allows the command to execute without
  // waiting for the entire cluster to converge to a single version
  // preventing weird deadlock situations. For instance, a transaction
  // with a schema change command might use a descriptor lease that is
  // at version: v - 1, and therefore deadlock when it tries to wait
  // for version: v, in the process of incrementing it to v + 1.
  // Therefore, a schema change command never increments the version,
  // and instead, sets the up_version boolean to notify the schema
  // changer execution engine that runs a future transaction to
  // increment the version.
  //
  // Multiple schema changes in the same transaction set up_version.
  // The actual schema change execution that follows a schema change
  // command sees the up_version boolean set, and increments the
  // table version after ensuring that there are no unexpired leases
  // for version - 1. The schema change execution must increment
  // the version before executing future state changes, to ensure
  // that the scheduled mutations made by the original commands are
  // visible on all leases. Multiple schema change mutations can be
  // grouped together on a particular version increment.

The cognitive burden of maintaining up_version has grown over time and it's time to
eliminate it.

This was put in primarily to not run into deadlocks, but a number of things have changed
 making eliminating the deadlock scenario easier.
1. commands that use a schema after it has been modified within the same transaction no longer use table leases, uncommittedTables within TableCollection.
2. A schema change when incrementing the version to v+1 can check that there is no descriptor at version v-1 in the TableCollection. If so, the entire transaction should be restarted, and only retried after waiting for all leases at version v-1 to be released (expired).

The other remaining question is how do we manage the two version lease invariant. i.e. We only permit leases on two versions of the table. This can be achieved by checking that there are no leases on version v-1 before incrementing the version v+1. Turns out this check is stronger than 2. mentioned above. Again, if a lease exists for v-1 the transaction must restart, wait for all leases at version v -1 before restarting.

**Comment by vivekmenezes** [comment]
@andreimatei raised a good question about whether we really care about retrying a transaction with timestamp T every time a transaction is unable to increment the version of a descriptor to v+1 . Can the transaction just wait until all leases on version v-1 have been released?

The problem is that another transaction in the system with timestamp T' > T can be using lease with version v-1. The transaction with timestamp T' can legitimately use the lease as long as v+1 has not been written with a timestamp <= T'

In other words the transaction has to push itself ahead of all releases/expiration time of leases on v-1
